### PR TITLE
Support Span<T> conversion

### PR DESCRIPTION
**Pull request by gukoff**
## Description

Support Span<T> conversion.

[**REVIEW THE FIRST COMMIT**](https://github.com/nanoframework/nanoFramework.IoT.Device/pull/29/commits/b4e68ffb2388fc784b037f3978585b4b090f81c4)

## Motivation and Context

No support existed before.

## How Has This Been Tested?

Verified regenerated projects work in Visual Studio.

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (work on Unit Tests, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/nanoFramework.IoT.Device.CodeConverter/Program.cs [hunk 1]
```diff
@@ -682,21 +691,24 @@ string outputDirectory
                     string line;
                     while (null != (line = input.ReadLine()))
                     {
-                        foreach (var replacement in replacements)
+                        // Replacing longer matches first is a safeguard heuristic.
```
Comment (gukoff) [inline src/nanoFramework.IoT.Device.CodeConverter/Program.cs:694]: Everything below this line is just refactoring that creeped into this PR

File: src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        /// <param name="length">The number of elements to include in the new System.Span</param>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// array is null, but start or length is non-zero. -or- start is outside the bounds
+        /// of the array. -or- start and length exceeds the number of elements in the array.
+        /// </exception>
+        public Span(T[] array, int start, int length)
+        {
+            if (array != null)
+            {
+                if ((start + length > array.Length) || (start >= array.Length))
```
Comment (Ellerbach) [inline src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs:39]: should be this: `if ((length > array.Length - start) || (start > array.Length))`

Check things here: https://github.com/nanoframework/CoreLibrary/blob/develop/nanoFramework.CoreLibrary/System/SpanByte.cs

File: src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        /// Gets the element at the specified zero-based index.
+        /// </summary>
+        /// <param name="index">The zero-based index of the element.</param>
+        /// <returns>The element at the specified index.</returns>
+        // public ref T this[int index] => ref _array[_start + index]; // <= this is not working and raises exception after few access
+        public T this[int index]
+        {
+            get
+            {
+                int realIndex = _start + index;
```
Comment (Ellerbach) [inline src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs:69]: the get should be:
```
if (index &gt; _length)                
{                    
throw new ArgumentOutOfRangeException();                }
}
return _array[_start + index];
```

File: src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+                {
+                    throw new ArgumentOutOfRangeException();
+                }
+
+                return _array[realIndex];
+            }
+
+            set
+            {
+                int realIndex = _start + index;
```
Comment (Ellerbach) [inline src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs:81]: set should be:
```
if (index > _length)
{
throw new ArgumentOutOfRangeException();
}
 _array[_start + index] = value;
```

File: src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+        /// <summary>
+        /// Forms a slice out of the current span that begins at a specified index.
+        /// </summary>
+        /// <param name="start">The index at which to begin the slice.</param>
+        /// <returns>A span that consists of all elements of the current span from start to the end of the span.</returns>
+        /// <exception cref="System.ArgumentOutOfRangeException">start is less than zero or greater than System.Span.Length.</exception>
+        public Span<T> Slice(int start)
+        {
+            return Slice(start, _length - start);
```
Comment (Ellerbach) [inline src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs:136]: missing the check for the length
```
if ((start > _length) || (start < 0))
{
throw new ArgumentOutOfRangeException();
}

return new SpanByte(_array, _start + start, _length - start);
```

File: src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        public static implicit operator Span<T>(T[] array)
+        {
+            return new(array);
+        }
+
+        /// <summary>
+        /// Defines an implicit conversion of a System.Span to a System.ReadOnlySpan.
+        /// </summary>
+        /// <param name="span">The object to convert to a System.ReadOnlySpan.</param>
+        public static implicit operator ReadOnlySpan<T>(Span<T> span)
```
Comment (Ellerbach) [inline src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs:181]: remove this one, we don't need ReadOnlySpan, we've simplify and get read of the ReadOnlySpan to only have Span

**Review by Ellerbach — COMMENTED** [review]
Please see my comments.
And please check the template here: https://github.com/nanoframework/CoreLibrary/blob/develop/nanoFramework.CoreLibrary/System/SpanByte.cs

File: src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs (hunk 2 above)
Comment (gukoff) [reply src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs:39]: Can start be equal to length, unless length=0?

File: src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs (hunk 3 above)
Comment (gukoff) [reply src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs:69]: Ah, I see. Length is not length of the array.

File: src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs (hunk 5 above)
Comment (gukoff) [reply src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs:136]: It should check it in the "main" metod `Slice(start, length)`

File: src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs (hunk 3 above)
Comment (gukoff) [reply src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs:69]: By the way it's `>=`, not `>`

File: src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs (hunk 2 above)
Comment (gukoff) [reply src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs:39]: This looks correct:
```
if ((start + length > array.Length) || (start > array.Length) || ( start == array.Length && array.Length > 0))
```

**Comment by gukoff** [comment]
If I see it right, SpanByte in mscorlib [currently](https://github.com/nanoframework/CoreLibrary/blob/9ae6c4be06ff5feebd869da9f4bd2ec72e411b05/nanoFramework.CoreLibrary/System/SpanByte.cs) has a bug and allows access 1 cell beyond the available length.

If the version in this PR looks good, I can move it to core library.

File: src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs (hunk 3 above)
Comment (Ellerbach) [reply src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs]: Good catch for this one. Mind a PR on msborlib? :-)

File: src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs (hunk 5 above)
Comment (Ellerbach) [reply src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs:135]: fair

**Review by Ellerbach — APPROVED** [review]

**Comment by Ellerbach** [comment]
@gukoff thanks! do you mind resolving the conflicts so we can merge this one soon?

**Pull request merged**